### PR TITLE
[14.0][ADD] base_generate_code module

### PR DESCRIPTION
**Pull request by Kev-Roche**
This module with a mixin model,  can be plugged to other models, in order to generate random codes  (e.g. for customers' purposes : coupon code, Gift card code, etc.).

It's inspired by <https://github.com/akretion/sale-promotion/tree/14.0-sale-coupon-custom-code-format/sale_coupon_custom_code_format> from @paradoxxxzero

Two questions : 
1) It is the good place for this module? 
2) The test include a fake model, maybe in a second time, split the tests in a second module for test only ?

cc @hparfr, @PierrickBrun

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Why not removing the word `customer` from the module name? Or is it linked to customers somehow?

**Comment by Kev-Roche** [comment]
> Why not removing the word `customer` from the module name? Or is it linked to customers somehow?

Indeed, the module, in some ways, can be usefull for partners in general or users.

File: base_generate_code/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+This module can be plugged to other models, in order to generate random codes (e.g. for customer's purposes : coupon code, Gift card code, etc.).
```
Comment (hparfr) [inline base_generate_code/readme/DESCRIPTION.rst]: The goal is to have human readable code: not too long, random enough, without some ambiguous char like ( "O" "0", "I", "L"), not memorable, without meaning (no words), acceptable as a barcode (like code128).

File: base_generate_code/models/code_format_mixin.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            while len(self.search([("code", "=", code)])):
+                code = rec._generate_code_from_mask(choices, mask)
+                retries += 1
+                if retries > self._dedup_max_retries:
+                    raise ValidationError(
+                        _("Unable to generate a non existing random code.")
+                    )
+            return code
+
+    code = fields.Char(compute=_generate_code, readonly=True, store=True)
```
Comment (hparfr) [inline base_generate_code/models/code_format_mixin.py]: fields are in the top  usually

**Review by hparfr — CHANGES_REQUESTED** [review]
You may update the readme to more reflects the purpose of the module.

File: base_generate_code/readme/DESCRIPTION.rst (hunk 1 above)
Comment (Kev-Roche) [reply base_generate_code/readme/DESCRIPTION.rst]: good idea, it's done.

**Comment by Kev-Roche** [comment]
@pedrobaeza , @PierrickBrun what do you think of the failed runboat/build here?

**Comment by pedrobaeza** [comment]
https://github.com/sbidoul/runboat/issues/48

**Comment by Kev-Roche** [comment]
> [sbidoul/runboat#48](https://github.com/sbidoul/runboat/issues/48)

Thank you!

**Review by hparfr — APPROVED** [review]

**Review by rousseldenis — APPROVED** [review]
LGTM

**Comment by sebastienbeau** [comment]
desolé mon petit bot pour l'erreur de frappe : /ocabot merge patch

File: base_generate_code/models/code_format_mixin.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                char
+                for char in ascii_uppercase
+                if char not in self._forbidden_characters
+            ],
+            "x": [
+                char
+                for char in ascii_lowercase
+                if char not in self._forbidden_characters
+            ],
+            "0": digits,
```
Comment (petrus-v) [inline base_generate_code/models/code_format_mixin.py:88]: I was just given an eye on this module, I suspect that `0` can be part of the final result as there is no filtering chars here like for ascii_uppercase and ascii_lowercase. I'm wondering why ?

**Pull request merged**